peerprep-backend/src/main/java/com/peerprep/peerprepbackend/filter/JwtAuthFilter.java

```diff
@@ -0,0 +1,65 @@
+package com.peerprep.peerprepbackend.filter;
+
+import com.auth0.jwt.JWT;
+import com.auth0.jwt.JWTVerifier;
+import com.auth0.jwt.algorithms.Algorithm;
+import com.auth0.jwt.exceptions.JWTVerificationException;
+import com.auth0.jwt.interfaces.DecodedJWT;
+import jakarta.servlet.FilterChain;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.lang.NonNull;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.web.authentication.WebAuthenticationDetails;
+import org.springframework.stereotype.Component;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+import java.io.IOException;
+
+@RequiredArgsConstructor
+@Component
+public class JwtAuthFilter extends OncePerRequestFilter {
+
+    @Value("${jwt.secret}")
+    private String jwtSecret;
+
+    @Value("${spring.application.name}")
+    private String appName;
+
+    private final UserDetailsService userDetailsService;
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, @NonNull HttpServletResponse response, @NonNull FilterChain filterChain) throws ServletException, IOException {
+        final String bearerToken = request.getHeader("Authorization");
+        if (bearerToken == null || !bearerToken.startsWith("Bearer")) {
+            filterChain.doFilter(request, response);
+            return;
+        }
+        final String jwt = bearerToken.substring(7);
+        DecodedJWT decodedJWT;
+        try {
+            Algorithm algorithm = Algorithm.HMAC256(jwtSecret);
+            JWTVerifier verifier = JWT.require(algorithm)
+                    .withIssuer(appName)
+                    .build();
+            decodedJWT = verifier.verify(jwt);
+        } catch (JWTVerificationException e) {
+            response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+            response.getWriter().write("Invalid or expired JWT token");
+            filterChain.doFilter(request, response);
+            return;
+        }
+        String username = decodedJWT.getSubject();
+        UserDetails user = userDetailsService.loadUserByUsername(username);
```

Comment:
would it be possible not to hit the db every time we check a request? rationale behind is this:

<img width="272" alt="image" src="https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g49/assets/70307497/2b07eb24-de64-491f-b035-9de94d8213a2">

because once we verify the jwt alr, it means that the jwt is legit and we don't really need to specifically retrieve user info. feels like quite a bit of overhead to alw hit the db for every request and i think it reduces the statelessness property of jwt?

<img width="511" alt="image" src="https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g49/assets/70307497/ca34404b-8d00-4595-a129-e2da74c4f1ab">
<br/><br/>
 a possible code could look like this?

```
String username = decodedJWT.getSubject();
List<GrantedAuthority> authorities = new ArrayList<>();
// Extract user roles or authorities from JWT and populate the 'authorities' list.

UsernamePasswordAuthenticationToken authToken = new UsernamePasswordAuthenticationToken(username, null, authorities);
authToken.setDetails(new WebAuthenticationDetails(request));
SecurityContextHolder.getContext().setAuthentication(authToken);
```

not very sure though so would really like your input on this :3

Reply:
that is valid... i think thats the more correct way to do it. but if the user's role changes, then we would need to issue a new jwt and blacklist the old one. since i haven't set up any method for manually invalidating tokens (i don't think this should be high prioirty?) so i think this way would be ok (for now)

Reply:
yep LGTM, i think we change only if it's really slowing down the calls, otherwise can just leave it?